File: 프로그래머스/lv2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.py
def dfs(start, visited, graph):
    visited[start] = True
    
    for i in range(1, len(graph)):
        if graph[start][i] == 1:
            if not visited[i]:
                dfs(i, visited, graph)
                    
def solution(n, wires):
    graph = [[0] * (n+1) for _ in range(n+1)]
    
    for a, b in wires:
        graph[a][b] = 1
        graph[b][a] = 1
    
    result = 1e9
    
    for a, b in wires:
        graph[a][b] = 0
        graph[b][a] = 0
        
        visited = [False] * (n+1)
        
        temp = []
        arr = []
        for i in range(1, n+1):
            if not visited[i]:
                dfs(i, visited, graph)
                arr.append(i)
                temp.append(visited.count(True))
        
        if len(temp) > 1:
            temp[1] = temp[1] - temp[0]
            result = min(result, abs(temp[1] - temp[0]))
            
        graph[a][b] = 1
        graph[b][a] = 1
    
    return result
        
        
                

File: 프로그래머스/2/86971. 전력망을 둘로 나누기/전력망을 둘로 나누기.java
import java.util.*;

class Solution {
    List<List<Integer>> graph;
    int N;
    public int solution(int n, int[][] wires) {
        graph = new ArrayList<>();
        N = n;
        
        for(int i = 0; i <= n; i++){
            graph.add(new ArrayList<>());    
        }
        
        for(int i = 0; i < N - 1; i++){
            int a = wires[i][0];
            int b = wires[i][1];
            
            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        
        int result = Integer.MAX_VALUE;
        
        for(int i = 0; i < N - 1; i++){
            int a = wires[i][0];
            int b = wires[i][1];
            
            graph.get(a).remove(graph.get(a).indexOf(b));
            graph.get(b).remove(graph.get(b).indexOf(a));
            
            boolean[] visited = new boolean[N + 1];
            int[] temp = new int[2];
            int idx = 0;
            
            for(int j = 1; j <= N; j++){
                if(!visited[j]){
                    temp[idx++] = dfs(j, visited);
                }
            }
            
            result = Math.min(result, Math.abs(temp[0] - temp[1]));
            
            graph.get(a).add(b);
            graph.get(b).add(a);
        }
        
        return result;
    }
    
    public int dfs(int start, boolean[] visited){
        visited[start] = true;
        
        int result = 1;
        
        for(int node : graph.get(start)){
            if(!visited[node]){
                result += dfs(node, visited);
            }
        }
        
        return result;
    }
}